src/reflect/scala/reflect/internal/tpe/GlbLubs.scala

```diff
@@ -230,14 +230,17 @@ private[internal] trait GlbLubs {
   def weakLub(tps: List[Type]): Type = (
     if (tps.isEmpty)
       NothingTpe
-    else if (tps forall isNumericValueType)
+    else if (shouldUseNumericLub(tps))
       numericLub(tps)
     else if (tps exists typeHasAnnotations)
       annotationsLub(lub(tps map (_.withoutAnnotations)), tps)
     else
       lub(tps)
   )
 
+  private def shouldUseNumericLub(tps: List[Type]) =
+    tps forall (tp => isNumericValueType(tp.widen))
```

Comment:
So I went and rebased this on top of the `-Yliteral-types` PR...

```
> scala -Xexperimental
[info] Running scala.tools.nsc.MainGenericRunner -usejavacp -Xexperimental
Welcome to Scala 2.12.4-20170731-141836-2899d86 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_121).
Type in expressions for evaluation. Or try :help.

scala> type I1 = 1
defined type alias I1

scala> type L1 = 1L
defined type alias L1

scala> def foo1(a: Option[Any]) = (a : @unchecked) match { case Some(x @ (_ : I1 | _ : L1)) => x }
foo1: (a: Option[Any])AnyVal

scala>
```

which is of course remedied by a well placed `.dealiasWiden` in `isNumericValueType`, but...

Neither `.widen` nor `.deconst` will make that work, and I suppose it's a question whether we expect it to at all. `.dealiasWiden` works, and doesn't seem to blow anything up, but it seems slightly dubious and probably approaching reason to rename `isNumericValueType`.